core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -742,7 +792,9 @@ private[deploy] class Master(
       while (numWorkersVisited < numWorkersAlive && !launched) {
         val worker = shuffledAliveWorkers(curPos)
         numWorkersVisited += 1
-        if (worker.memoryFree >= driver.desc.mem && worker.coresFree >= driver.desc.cores) {
+        if (canLaunchDriver(worker, driver.desc)) {
+          val allocated = worker.acquireResources(driver.desc.resourceReqs)
+          driver.withResources(allocated)
```

Comment:
I'm assuming driver could also not launch if worker doesn't have any GPUs and it requests them